src/pypromice/process/L1toL2.py

```diff
@@ -141,6 +156,248 @@ def toL2(L1, T_0=273.15, ews=1013.246, ei0=6.1071, eps_overcast=1.,
+         var_all = [k + '_u',k + '_l',k + '_i'] # apply to upper, lower boom, and instant
+         static_lim = var_threshold[k]['static_limit'] # loading static limit
+         diff_h = var_threshold[k]['diff_period'] # loading diff period
+         
+         for v in var_all:
+             if v in df:
+                
+                data = df[v]
+                diff = data.diff()
+                diff.fillna(method='ffill', inplace=True) # forward filling all NaNs! 
+                diff = np.array(diff)
+
+                diff_period = np.ones_like(diff) * False
+                
+                for i,d in enumerate(diff): # algorithm that ensures values can stay the same within the diff_period
+                    if i > (diff_h-1): 
+                        if sum(abs(diff[i-diff_h:i])) < static_lim:
+                            diff_period[i-diff_h:i] = True
+                
+                diff_period = np.array(diff_period).astype('bool')
+                
+                outliers = df[v][diff_period] # finding outliers in dataframe
+                
+                df.loc[outliers.index,v] = np.nan # setting outliers to NaN
+                
+    # Back to xarray, and re-assign the original attrs
+    ds_out = df.to_xarray()
+    ds_out = ds_out.assign_attrs(ds.attrs) # Dataset attrs
+    for x in ds_out.data_vars: # variable-specific attrs
+        ds_out[x].attrs = ds[x].attrs
+    # equivalent to above:
+    # vals = [xr.DataArray(data=df_out[c], dims=['time'], coords={'time':df_out.index}, attrs=ds[c].attrs) for c in df_out.columns]
+    # ds_out = xr.Dataset(dict(zip(df_out.columns, vals)), attrs=ds.attrs)
+    return ds_out
+    
+
+def percentileQC(ds):
+    '''
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Level 1 dataset
+
+    Returns
+    -------
+    ds_out : xr.Dataset
+        Level 1 dataset with percentile outliers set to NaN
+    '''
+    # Check if on-disk sqlite db exists
+    # If it not exists, then run compute_percentiles.py
+    
+    base_path = os.getcwd()
+    
+    file_path1 =  base_path + '/main/src/pypromice/qc/percentiles.db'
+    file_path2 =  base_path + '/qc/percentiles.db'
+   
+    script_path1 = base_path + '/main/src/pypromice/qc/compute_percentiles.py'
+    script_path2 = base_path + '/qc/compute_percentiles.py'
+    
+    script_exist1 = os.path.isfile(script_path1)
+    
+    db_exist1 = os.path.isfile(file_path1)
+    db_exist2 = os.path.isfile(file_path2)
+
+    
+    print(f'Does percentiles.db exist {db_exist2} in this path {file_path2}')    
+    
+    
+    if not db_exist1 and not db_exist2:
+        if script_exist1: 
+            print(f'percentiles.db does not exist running {script_path1}')
+            subprocess.call(['python',script_path1])
+            file_path = file_path1
+        else:
+            print(f'percentiles.db does not exist running {script_path2}')
+            subprocess.call(['python',script_path2])
+            file_path = file_path2
```

Comment:
@ladsmund I wrote this original code that uses sqlite. The general idea is to provide processing effficiency. If the percentile distributions (for each variable) are stored in a sql database, we are able to extract the percentiles of interest very fast with a sql query, rather than opening/reading a static file for each station. sqlite is the simplest database solution, where we can have a small file on disk.

However, this is of more significant benefit for much larger datasets (e.g. 10s or 100s of thousands of stations). A solution using a static file could also be implemented if desired. Either way, you are persisting something on disk between runs that holds the percentile distributions.